trackintel/preprocessing/positionfixes.py

```diff
@@ -173,187 +177,335 @@ def generate_triplegs(positionfixes, staypoints=None, *args, **kwargs):
+    user_id_this:
+
+    Returns
+    --------
+    list
+        a list of dictionaries with individual triplegs
+
+    """
+    generated_triplegs_list = []
+    stps = staypoints.loc[staypoints['user_id'] == user_id_this].sort_values('started_at')
+    stps = stps.reset_index().to_dict('records')
+    positionfixes = positionfixes.sort_values('tracked_at')
+    for stp1, stp2 in zip(list(stps), list(stps)[1:]):
+        # - Get all positionfixes that lie between these two staypoints by comparing timestamps.
+        # - generate tripleg
+
+        # Not so efficient, always matching on the time (as things are sorted anyways).
+        pfs_tripleg = positionfixes[(stp1['finished_at'] <= positionfixes['tracked_at']) &
+                                    (positionfixes['tracked_at'] <= stp2['started_at'])]
+        generated_triplegs_list.append(_get_tripleg_record_from_psfs(pfs_tripleg, user_id_this, min_nb_of_points=3))
+
+    # add first tripleg
+    pfs_first_tripleg = positionfixes[positionfixes['tracked_at'] <= stps[0]['started_at']]
+    generated_triplegs_list = [_get_tripleg_record_from_psfs(pfs_first_tripleg, user_id_this, min_nb_of_points=2
+                                                             )] + generated_triplegs_list
+
+    # add last tripleg
+    pfs_first_tripleg = positionfixes[positionfixes['tracked_at'] >= stps[-1]['finished_at']]
+    generated_triplegs_list.append(_get_tripleg_record_from_psfs(pfs_first_tripleg, user_id_this, min_nb_of_points=2))
+
+    # filter None values
+    return list(filter(None, generated_triplegs_list))
+
+def _triplegs_between_staypoints_case3(positionfixes, user_id_this):
+    """
+    This function uses column 'staypoint_id' to identify all positionfixes that lie in between two staypoints.
+
+    Parameters
+    ----------
+    positionfixes: trackintel positionfixes
+    user_id_this:
+
+    Returns
+    --------
+    list
+        a list of dictionaries with individual triplegs
+    """
+
+    name_geocol = positionfixes.geometry.name
+    generated_triplegs = []
+    # initialize first tripleg
+    prev_pf = None
+    prev_idx = None
+    curr_tripleg = {
+        'user_id': user_id_this,
+        'started_at': positionfixes['tracked_at'].iloc[0],
+        'finished_at': None,
+        'geom': [],
+        'pfs_ids': []
+    }
+
+    for idx, pf in positionfixes.iterrows():
+
+        # - loop through all positionfixes
+        # - identify the current situation and define the variable 'status'
+        # - store or skip the current positionfix based on the state of 'status'
+
+        # During the loop the status of a positionfix can be {'in_tripleg', 'in_staypoint', 'tripleg_starts',
+        # 'tripleg_ends'}
+        if prev_pf is None and np.isnan(pf['staypoint_id']):
+            status = 'in_tripleg'
+        elif prev_pf is None and ~np.isnan(pf['staypoint_id']):
+            status = 'in_staypoint'
+
+        elif ~np.isnan(prev_pf['staypoint_id']):
```

Comment:
I am not able to understand what happened here as the logic is quite complicated - -!. But Vscode told me that `Value 'prev_pf' is unsubscriptable`, maybe it's because prev_pf is initialized as None and only assigned as the previous pf in the first loop (at the end). It might not cause an error, but I think it is not safe in later stages. Is there a way to fix this?

Reply:
I adjusted the code a bit but these state-machine like implementations are never really pretty (nor safe super but that's what tests are for^^). I have added an extra condition with an extra flag for the first iteration (that's what the None was for) but I guess that the Vscode will now complain about that the variable might be called before being defined. I thought about giving it some initialization but I did not have a good idea that does not make it more dangerous instead of more safe.